core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -439,6 +444,11 @@ private void initSession(SessionProtocol desiredProtocol, SerializationFormat se
         final EventLoop eventLoop = ch.eventLoop();
         assert eventLoop.inEventLoop();
 
+        int connectTimeoutMillis = this.connectTimeoutMillis;
+        if (poolKey.proxyConfig != ProxyConfig.direct()) {
+            // Add a bit more timeout to prevent a race between the proxy timeout and session timeout.
+            connectTimeoutMillis += 100;
```

Comment:
Would it be cleaner if we just rely on the proxy timeout only if forward proxy instead of setting an arbitrary value?

e.g.
```java
final ScheduledFuture<?> timeoutFuture;
if (poolKey.proxyConfig.proxyType().isForwardProxy()) {
    timeoutFuture = NOOP_FUTURE;
} else {
    timeoutFuture = eventLoop.schedule(() -> {
        if (sessionPromise.tryFailure(new SessionProtocolNegotiationException(
                desiredProtocol, "connection established, but session creation timed out: " + ch))) {
            ch.close();
        }
    }, connectTimeoutMillis, TimeUnit.MILLISECONDS);
}
```

Reply:
That is a brilliant idea. 👍

Reply:
On second thought, I think we still need `timeoutFuture` because `ProxyHandler` handles the timeout until `ProxyConnectionEvent` is published but a session protocol is negotiated after that.

Reply:
Or alternatively, I guess we can just throw a proxy connect exception explicitly.

We may need to remove the `maybeHandleProxyFailure` call since `HttpSessionHandler` will handle it
```java
diff --git a/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java b/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java
index 2d02e1bed..6b857d937 100644
--- a/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java
+++ b/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java
@@ -445,13 +445,15 @@ final class HttpChannelPool implements AsyncCloseable {
         assert eventLoop.inEventLoop();
 
         int connectTimeoutMillis = this.connectTimeoutMillis;
-        if (poolKey.proxyConfig != ProxyConfig.direct()) {
-            // Add a bit more timeout to prevent a race between the proxy timeout and session timeout.
-            connectTimeoutMillis += 100;
-        }
         final ScheduledFuture<?> timeoutFuture = eventLoop.schedule(() -> {
-            if (sessionPromise.tryFailure(new SessionProtocolNegotiationException(
-                    desiredProtocol, "connection established, but session creation timed out: " + ch))) {
+            final Exception exception;
+            if (poolKey.proxyConfig.proxyType().isForwardProxy()) {
+                exception = new ProxyConnectException();
+            } else {
+                exception = new SessionProtocolNegotiationException(
+                        desiredProtocol, "connection established, but session creation timed out: " + ch);
+            }
+            if (sessionPromise.tryFailure(exception)) {
                 ch.close();
             }
         }, connectTimeoutMillis, TimeUnit.MILLISECONDS);
@@ -538,9 +540,6 @@ final class HttpChannelPool implements AsyncCloseable {
                 });
             } else {
                 final Throwable throwable = future.cause();
-                if (throwable instanceof ProxyConnectException) {
-                    maybeHandleProxyFailure(desiredProtocol, key, throwable);
-                }
                 promise.completeExceptionally(UnprocessedRequestException.of(throwable));
             }
         } catch (Exception e) {
```

Reply:
In your suggestion, `ProxyConnectException` could be raised even though the proxy connection is established.

What do you think of scheduling the session timeout when `ProxyConnectionEvent` is received? The maximum connection time for a proxy connection could be doubled which, I think, would make sense because two connections will sequentially be made.
https://github.com/line/armeria/blob/3822432319e5ccc4c6a4d399abc08fca9e98caaa/core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java#L440-L441

Reply:
> What do you think of scheduling the session timeout when ProxyConnectionEvent is received? 

That also sounds like a good idea 👍